File: LinkedList.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace LinkedList
{
    public class MyLinkedList<TNode>
    {
        private Node<TNode> Head;

        private class Node<TNode>
        {
            public TNode Data;
            public Node<TNode> Right { get; set; }

            public Node(TNode data)
            {
                Data = data;
            }

            public override string ToString()
            {
                return Data.ToString();
            }
        }

        public void SetNewHead(TNode data)
        {
            if (Head == null)
            {
                Head = new Node<TNode>(data);
            }
            else
            {
                Node<TNode> NewHead = new Node<TNode>(data);
                NewHead.Right = Head;
                Head = NewHead;
            }
        }

        public void AddItem(TNode data)
        {
            if (Head == null)
            {
                SetNewHead(data);
            }
            else
            {
                Node<TNode> Current = Head;
                while (Current.Right != null)
                {
                    Current = Current.Right;
                }
                Current.Right = new Node<TNode>(data);
            }
        }

        public void AddAtIndex(TNode data, int index)
        {
            if (index == 0)
            {
                SetNewHead(data);
            }
            else
            {
                Node<TNode> Current = Head;
                int count = 0;
                while (Current != null)
                {
                    if ((count + 1) == index)
                    {
                        Node<TNode> NewNode = new Node<TNode>(data);
                        NewNode.Right = Current.Right;
                        Current.Right = NewNode;
                    }
                    Current = Current.Right;
                    count++;
                }
            }
        }

        public void DeleteItem(TNode data)
        {
            if (Head.Data.Equals(data))
            {
                Head = Head.Right;
            }

            Node<TNode> Current = Head;
            while (Current != null)
            {
                if (Current.Right.Data.Equals(data))
                {
                    Current.Right = Current.Right.Right;
                }
                Current = Current.Right;
            }
        }

        public void DeleteByIndex(int index)
        {
            Node<TNode> Current = Head;
            int count = 0;
            if (index == 0)
            {
                DeleteItem(Head.Data);
            }
            while (Current != null)
            {
                if (index == count)
                {
                    DeleteItem(Current.Data);
                }
                Current = Current.Right;
                count++;
            }
        }

        public void SearchItem(int index)
        {
            Node<TNode> Current = Head;
            int count = 0;
            if (index == 0)
            {
                Console.WriteLine(Current.Data);
            }
            while (Current != null)
            {
                if (count == index)
                {
                    Console.WriteLine(Current.Data);
                }

                Current = Current.Right;
                count++;
            }
        }
    }
}

File: LinkedList.js
class Node {
  constructor(data) {(
        this.data = data,
        this.next = null,
        this.previous = null
        )}
}

class LinkedList {
  constructor() {
    this.head = null;
    this.tail = null;
  }

    setNewHead = data => {
        var newHead = new Node(data);
        if(this.head) {
            newHead.next = this.head;
            this.head.previous = newHead;
        }
        this.head = newHead;
    };

    addItem = (data, index = null) => {
        let newItem = new Node(data);
        if(index)
        {
            let current = this.searchItem(index);
            if(current) {
                current.previous.next = newItem;
                current.previous = newItem;
            }
            this.addItem(data);
        } else {
            if (this.head) {
                let current = this.head;
                while(current.next)
                {
                    current = current.next;
                }
                current.next = newItem;
                newItem.previous = current;
                this.tail = newItem;
            } else {
                this.setNewHead(data);
            }
        }
    };
    
    searchItem = index => {
        if (index == 0) {
            return this.head;
        } else {
            let count = 0;
            let current = this.head;
            while(current)
            {
                if (count == index) {
                    return current;
                }
                count++;
                current = current.next;
            }
        }
    };
    
    deleteItem = data => {
        if(data == this.head.data) {
            if (this.head.next) {
                this.head = this.head.next;
            } else {
                this.head = null;
            }
        } else {
            let current = this.head;
            while(current) {
                if (current.data == data)
                {
                    this.tail = current == this.tail ? current.previous : this.tail;
                    current.previous.next = current.next;
                }
                current = current.next;
            }
        }
    };
    
    deleteByIndex = index => {
        let toDelete = this.searchItem(index);
        this.deleteItem(toDelete.data);
        console.log(myList);
    };
    
};